### PR TITLE
Handled near cache compatibility for versions smaller than 3.8

### PR DESCRIPTION
**Pull request by ahmetmircik**
closes https://github.com/hazelcast/hazelcast/issues/11343

A latest [fix](https://github.com/hazelcast/hazelcast/pull/11003) was caused this regression. The near cache invalidation listener has 2 modes: one for the versions < 3.8 and one for the versions >= 3.8. 
In this PR, after getting version of connected server, we provide only one mode can work and other one stays untouched and doesn't try to work unnecessarily. 

As a result, we shouldn't see any noise in the logs due to the operation retries of an incompatible mode.

### CODE REVIEW

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java [hunk 1]
```diff
@@ -796,18 +780,24 @@ public void onFailure(Throwable t) {
...
+
         private volatile boolean supportsRepairableNearCache;
 
         @Override
         public void beforeListenerRegister() {
-            pre38EventHandler.beforeListenerRegister();
-            repairingEventHandler.beforeListenerRegister();
+            supportsRepairableNearCache = supportsRepairableNearCache();
+
+            if (supportsRepairableNearCache) {
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: Could be simplified to:
```java
            repairingEventHandler.beforeListenerRegister();
            if (!supportsRepairableNearCache) {
                // call repairingEventHandler here to clean possibly left resources.
                pre38EventHandler.beforeListenerRegister();
            }
```

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java [hunk 2]
```diff
@@ -546,13 +542,13 @@ private void registerInvalidationListener() {
             return;
         }
 
-        EventHandler eventHandler = createInvalidationEventHandler();
-        nearCacheMembershipRegistrationId = addNearCacheInvalidationListener(eventHandler);
+        ConnectedServerVersionAwareNearCacheEventHandler handler = new ConnectedServerVersionAwareNearCacheEventHandler();
+        nearCacheMembershipRegistrationId = addNearCacheInvalidationListener(handler);
+        if (!handler.supportsRepairableNearCache) {
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: The usage of the `supportsRepairableNearCache` field and method is a bit confusing for me (as in: when is the field initialized, when do I have to use the method instead etc.). But the classes and their hierarchy and lifecycle are really not easy to read on GitHub.

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java [hunk 3]
```diff
@@ -833,35 +823,37 @@ public void handle(ClientMessage clientMessage) {
...
         }
 
         @Override
         public void onListenerRegister() {
+            // NOP
         }
 
         @Override
         public void handle(String name, Data key, String sourceUuid, UUID partitionUuid, long sequence) {
+            assert repairingHandler != null;
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: Those assertions will create a code coverage branch on the class level, which needs to be tested. Since we'll just get an `AssertionError` instead of a `NPE`, I don't think this `assert` will have much benefit. It's not that there is a silent error will be prevented by it. Same in the other `handle()` method.

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java [hunk 4]
```diff
@@ -833,35 +823,37 @@ public void handle(ClientMessage clientMessage) {
...
+        private volatile RepairingHandler repairingHandler;
 
         @Override
         public void beforeListenerRegister() {
-
+            if (supportsRepairableNearCache()) {
+                RepairingTask repairingTask = getContext().getRepairingTask(CacheService.SERVICE_NAME);
+                repairingHandler = repairingTask.registerAndGetHandler(nameWithPrefix, nearCache);
+            } else {
+                RepairingTask repairingTask = getContext().getRepairingTask(MapService.SERVICE_NAME);
```
Comment (Donnerbart) [inline hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: `MapService` in a cache proxy looks wrong. Should this both just be `CacheService.SERVICE_NAME`?

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java (hunk 2 above)
Comment (ahmetmircik) [reply hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: fixed

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java (hunk 1 above)
Comment (ahmetmircik) [reply hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: fixed

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java (hunk 4 above)
Comment (ahmetmircik) [reply hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: very good finding. fixed

File: hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java (hunk 3 above)
Comment (ahmetmircik) [reply hazelcast-client/src/main/java/com/hazelcast/client/cache/impl/NearCachedClientCacheProxy.java]: fixed

File: hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java [hunk 5]
```diff
@@ -101,6 +105,17 @@ protected final HazelcastClientInstanceImpl getClient() {
         return (HazelcastClientInstanceImpl) getContext().getHazelcastInstance();
     }
 
+    protected final int getConnectedServerVersion() {
+        HazelcastClientInstanceImpl client = getClient();
+        ClientConnectionManager connectionManager = client.getConnectionManager();
+        Collection<ClientConnection> activeConnections = connectionManager.getActiveConnections();
```
Comment (vbekiaris) [inline hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java]: Previously we were checking the server version on the owner connection. Why was this changed to lookup server version on any active connection?

Comment (ahmetmircik) [reply hazelcast-client/src/main/java/com/hazelcast/client/spi/ClientProxy.java]: Actually it is not required to be owner. On client side, even if you have a healthy connection it is not guaranteed that owner is decided even on a one node cluster. Decision of an owner connection is another parallel process. (Maybe this is a bug but this is the current situation) So using one of existing connection should be ok.
